openwisp_controller/connection/tests/test_models.py

```diff
@@ -321,6 +321,14 @@ def test_ssh_exec_exception(self, *args):
             dc.connector_instance.disconnect()
             mocked.assert_called_once()
 
+    @mock.patch('openwisp_controller.connection.models.DeviceConnection.send_working_status_changed_signal')
+    def test_working_status_change_signal_sent(self, mock):
+        ckey = self._create_credentials_with_key(port=self.ssh_server.port)
+        dc = self._create_device_connection(credentials=ckey)
+        dc.is_working = True
+        dc.save()
+        self.assertTrue(mock.called)
```

Comment:
Look at:

- https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/monitoring/tests/test_models.py#L181-L233
- https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/tests.py

It's a nice solution that I found on the internet a while ago (I think stackoverflow).
We can then move that test utility to openwisp-utils to avoid duplicating it in several modules: https://github.com/openwisp/openwisp-utils/tree/master/openwisp_utils
That's how we do things in OpenWISP. I will create the issue for that later, in the meanwhile you can duplicate the code.

Reply:
Thanks for the review @nemesisdesign, I have updated the PR. I had to use `test_signal` (as the name for file containing `catch_signal` function) rather than `tests` due to presence of directory of the same name `tests` (using `tests` as name leads to **import error** which is expected) but the name appears to be a bit misleading and vague to me, `tests` was more generic name. Also, I found it appropriate to create the file in `openwisp_controller` directory as it can be used by other apps too besides `connection`. So should I rename it (if so any suggestions) or leave it right now as it would be moved to openwisp-utils later on :smile: